File: NotiMemoService.java
package com.example.notimemo;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.Service;
import android.content.Intent;
import android.os.Build;
import android.os.IBinder;

import androidx.annotation.Nullable;
import androidx.core.app.NotificationCompat;

// Foreground Service: 앱이 종료되어도 알림을 계속 유지하기 위한 서비스 클래스
public class NotiMemoService extends Service {

    private static final String CHANNEL_ID = "notimemo_channel"; // 알림 채널 ID

    // 서비스가 시작될 때 호출되는 함수
    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        // 메모 내용 전달받기
        String memo = intent.getStringExtra("memo");

        // 알림 채널 만들기 (API 26+)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel channel = new NotificationChannel(
                    CHANNEL_ID,
                    "NotiMemo Channel",
                    NotificationManager.IMPORTANCE_DEFAULT
            );
            channel.setDescription("메모 고정 알림");

            NotificationManager manager = getSystemService(NotificationManager.class);
            if (manager != null) {
                manager.createNotificationChannel(channel);
            }
        }

        // 알림 구성
        Notification notification = new NotificationCompat.Builder(this, CHANNEL_ID)
                .setContentTitle("NotiMemo")
                .setContentText(memo != null ? memo : "메모 없음")
                .setSmallIcon(R.drawable.ic_launcher_foreground)
                .setOngoing(true)
                .build();

        // Foreground Service 시작
        startForeground(1, notification);

        // 시스템이 종료해도 자동 재시작
        return START_STICKY;
    }
    // 바인딩 기능은 사용하지 않음.
    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }
}

File: android/app/src/main/kotlin/com/example/notimemo/NotiMemoService.kt
package com.example.notimemo

import android.app.*
import android.content.Context
import android.content.Intent
import android.content.pm.ServiceInfo
import android.graphics.Typeface
import android.os.Build
import android.os.IBinder
import android.text.SpannableString
import android.text.style.StyleSpan
import androidx.core.app.NotificationCompat
import androidx.core.app.ServiceCompat

class NotiMemoService : Service() {
    companion object {
        const val CHANNEL_ID = "notimemo_fg_channel"
        const val NOTIFICATION_ID = 1
        const val ACTION_STOP = "com.example.notimemo.STOP"
        const val ACTION_REPOST = "com.example.notimemo.REPOST"
        const val EXTRA_MEMO = "memo"
        private const val PREFS_NAME = "notimemo_prefs"
        private const val PREF_MEMO = "current_memo"
        private const val PREF_TIME = "created_time"
    }

    override fun onBind(intent: Intent?): IBinder? = null

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        when (intent?.action) {
            ACTION_STOP -> {
                clearSavedMemo()
                getSharedPreferences("FlutterSharedPreferences", Context.MODE_PRIVATE)
                    .edit()
                    .remove("flutter.saved_memo")
                    .remove("flutter.notification_active")
                    .apply()
                sendBroadcast(Intent("com.example.notimemo.DISMISSED").apply {
                    setPackage(packageName)
                })
                ServiceCompat.stopForeground(this, ServiceCompat.STOP_FOREGROUND_REMOVE)
                stopSelf()
                return START_NOT_STICKY
            }
            ACTION_REPOST -> {
                val memo = intent.getStringExtra(EXTRA_MEMO) ?: getSavedMemo() ?: return START_NOT_STICKY
                showNotification(memo, getSavedTime())
                return START_STICKY
            }
            else -> {
                val memo = intent?.getStringExtra(EXTRA_MEMO) ?: getSavedMemo() ?: return START_NOT_STICKY
                val now = System.currentTimeMillis()
                saveMemo(memo, now)
                showNotification(memo, now)
                return START_STICKY
            }
        }
    }

    private fun showNotification(memo: String, createdAt: Long) {
        ensureChannel()

        val stopIntent = PendingIntent.getService(
            this, 0,
            Intent(this, NotiMemoService::class.java).apply { action = ACTION_STOP },
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )

        val repostIntent = PendingIntent.getService(
            this, 2,
            Intent(this, NotiMemoService::class.java).apply {
                action = ACTION_REPOST
                putExtra(EXTRA_MEMO, memo)
            },
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )

        val boldMemo = SpannableString(memo).apply {
            setSpan(StyleSpan(Typeface.BOLD), 0, memo.length, 0)
        }

        val notification = NotificationCompat.Builder(this, CHANNEL_ID)
            .setSmallIcon(R.drawable.ic_notify_white)
            .setContentText(boldMemo)
            .setOngoing(true)
            .setSilent(true)
            .setWhen(createdAt)
            .setShowWhen(true)
            .setPriority(NotificationCompat.PRIORITY_LOW)
            .addAction(0, "지우기", stopIntent)
            .setDeleteIntent(repostIntent)
            .build()

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            startForeground(NOTIFICATION_ID, notification, ServiceInfo.FOREGROUND_SERVICE_TYPE_MEDIA_PLAYBACK)
        } else {
            startForeground(NOTIFICATION_ID, notification)
        }
    }

    private fun saveMemo(memo: String, time: Long) {
        getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE).edit()
            .putString(PREF_MEMO, memo)
            .putLong(PREF_TIME, time)
            .apply()
    }

    private fun getSavedTime(): Long =
        getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            .getLong(PREF_TIME, System.currentTimeMillis())

    private fun getSavedMemo(): String? =
        getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE).getString(PREF_MEMO, null)

    private fun clearSavedMemo() {
        getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE).edit()
            .remove(PREF_MEMO)
            .remove(PREF_TIME)
            .apply()
    }

    private fun ensureChannel() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) return
        val nm = getSystemService(NotificationManager::class.java)
        if (nm.getNotificationChannel(CHANNEL_ID) != null) return
        val channel = NotificationChannel(CHANNEL_ID, "알림메모", NotificationManager.IMPORTANCE_LOW).apply {
            description = "알림 메모를 표시합니다"
            setSound(null, null)
            enableLights(false)
            enableVibration(false)
        }
        nm.createNotificationChannel(channel)
    }
}
